Add missing events

zale144 opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v           ✰  Thanks for creating a PR! You're awesome! ✰
v Please note that maintainers will only review those PRs with a completed PR template.
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Purpose of Changes and their Description

## Link(s) to Ticket(s) or Issue(s) resolved by this PR

## Are these changes tested and documented?

- [x] If tested, please describe how. If not, why tests are not needed. - probably not needed to test event emissions
- [x] If documented, please describe where. If not, describe why docs are not needed. - not needed
- [x] Added to `Unreleased` section of `CHANGELOG.md`?

zale144 committed:
Add missing events

guilherme-brandao committed:
Added new events

amimart commented on x/emissions/proto/emissions/v9/events.proto:
```diff
@@ -62,6 +67,322 @@ message EventNetworkInferences {
+    (gogoproto.nullable) = false,
+    (amino.dont_omitempty) = true
+  ];
+  string topic_stake = 4 [
```
I'd remove that as we should be able to compute it by replaying stake related events

zale144 replied:
removed

xmariachi commented on x/emissions/proto/emissions/v9/events.proto:
```diff
@@ -3,7 +3,12 @@ package emissions.v9;
+import "emissions/v3/topic.proto";
+import "emissions/v8/params.proto";
 import "gogoproto/gogo.proto";
+import "amino/amino.proto";
```
Why these new protos are used for? 
And if they are used here, should they be used in other parts too for alignment?

zale144 replied:
they are used in other places as well. `amino` is used for when `(amino.dont_omitempty) = true` is needed

xmariachi commented on x/emissions/proto/emissions/v9/events.proto:
```diff
@@ -62,6 +67,322 @@ message EventNetworkInferences {
+  uint64 topic_id = 1;
+  string amount = 2 [
+    (cosmos_proto.scalar) = "cosmos.Int",
+    (gogoproto.customtype) = "cosmossdk.io/math.Int",
```
1. Why these new protos are used? If they are used here, should they be used in other parts too for alignment?

2. Since this is math.Int, I fear this could overflow - shouldn't our Dec be used instead?

zale144 replied:
it's the same type as the value being passed. check `(k *Keeper) AddTopicFeeRevenue`. I'd link it here, but the file is too big 😁

amimart replied:
Yeah and the cosmos Int type is 256bit so I'd expect the Dec to overflow before it

xmariachi replied:
ok, the same type should be used, I assumed it was receiving a Dec. 
(However Dec would not overflow before it, it has a `MaxExponent = 100000`... although our input layer's Dec only 40 iirc)

xmariachi reviewed (commented):
Lgtm in general, just two comments on protos

amimart commented on x/emissions/proto/emissions/v9/events.proto:
```diff
@@ -62,6 +67,322 @@ message EventNetworkInferences {
+    (amino.dont_omitempty) = true
+  ];
+  int64 block_removal_started = 4;
+  int64 block_removal_completed = 5;
```
I think this should be deduced

zale144 replied:
removed

amimart commented on x/emissions/proto/emissions/v9/events.proto:
```diff
@@ -62,6 +67,322 @@ message EventNetworkInferences {
+  emissions.v3.Topic topic = 1;
+}
+
+message EventAddTopicFeeRevenue {
```
If we already have events for topic funding, registration and payload submission I don't see the value of this one

zale144 replied:
not sure if you mean you see or you don't see the value of this one?

amimart replied:
I don't see the value of this one, I think we already have the needed info through other events, to be checked though

zale144 replied:
removed

amimart commented on x/emissions/proto/emissions/v9/events.proto:
```diff
@@ -62,6 +67,322 @@ message EventNetworkInferences {
 
+message EventInsertWorkerPayload {
+  uint64 topic_id = 1;
+  emissions.v3.WorkerDataBundle worker_data_bundle = 2;
```
Do we have an idea of how big data bundles can be? I feel like this could bring some overhead on the network as I've seen in the past some blocks with 5000 submission txs.

I'm always in favour of putting everything needed in events to recompose the state only relying on events, but this could be an exception.

I'm just putting a warning, not saying we shouldn't go with it, but running some benchmark on that could be justified

zale144 replied:
we already have NetworkInferences and NetworkLoss, which are I think even bigger 🤷‍♂️

amimart replied:
Yeah but these are emitted only once per epoch.

I think having the whole data bundle is not necessary, for example we don't need the signature and the public key in the event as at this point it is already verified. We only need the sending address, topic id/nonce and the inference value or forecast elements

zale144 replied:
reworked

amimart commented on x/emissions/proto/emissions/v9/events.proto:
```diff
@@ -62,6 +67,322 @@ message EventNetworkInferences {
+  int64 block_removal_completed = 5;
+}
+
+message EventCancelRemoveReputerStake {
```
Do we need events for when a wallet ask for stake removal or to cancel stake removal? To observe stake evolution I think we only need two events: when stake is added and when stake is effectively removed, I don't see the value tracking internal stake removal scheduling

zale144 replied:
removed

zale144 replied:
will be addressed in @guilherme-brandao's  PR

amimart commented on x/emissions/proto/emissions/v9/events.proto:
```diff
@@ -62,6 +67,322 @@ message EventNetworkInferences {
+  int64 block_removal_completed = 5;
+}
+
+message EventAddDelegateStake {
```
Currently we have two staking msg on reputers, one for the reputer itself and one for others but it's the same thing, I think we should have only one set of events for both. It's even easier to manage on the indexer side.

zale144 replied:
removed

amimart commented on x/emissions/proto/emissions/v9/events.proto:
```diff
@@ -62,6 +67,322 @@ message EventNetworkInferences {
+  string address = 2;
+}
+
+message EventTopicAdminWhitelistBulkAdded {
```
I think we should have a single event associated to a specific state mutation, bulk operations just run the same mutation multiple times which should emit multiple events, not a different one imho

zale144 replied:
so you would rather emit an event for each address added or removed, instead of one event per bulk? It might be a lot of events, not sure

amimart replied:
To be honest I don't think having bulk msg is a good practice as we can just put multiple message in a tx.
Maybe a compromise can be found with a single event type having a repeated field.

zale144 replied:
removed, done with events per address

amimart reviewed (commented):
Thanks!

I've made a few comments related to mainly two concerns:

1. We should dissociate state mutation and tx msg: different msgs can lead to the same mutation, but we shouldn't have an event for each one of them.

2. Events are not free: we should avoid unnecessary data, duplication and very large payloads.

These are only my opinion, I'd be happy to have yours and discuss that

zale144 committed:
Some fixes for PR comments

zale144 committed:
More fixes for PR comments

zale144 committed:
Merge remote-tracking branch 'origin/dev' into alek/add-missing-events

zale144 committed:
follow up

zale144 committed:
remove redundant event attribute field

zale144 committed:
further make events more compact

guilherme-brandao committed:
Added new events

guilherme-brandao committed:
Merge remote-tracking branch 'origin/alek/add-missing-events' into guilherme/engn-4091-add-missing-events

guilherme-brandao committed:
Update pb and pulsar

guilherme-brandao committed:
Fix lint

guilherme-brandao committed:
Fix CI

guilherme-brandao committed:
Fix CI

guilherme-brandao committed:
Fixes after reviews

guilherme-brandao committed:
Added new event in for the openning of the reputer window

guilherme-brandao committed:
Added individual topic status events and fixed reputer window event

guilherme-brandao committed:
Updated topic weight event and topic query

guilherme-brandao committed:
Remove window start bock from events

guilherme-brandao committed:
Update topic status implementation

guilherme-brandao committed:
Revert "Update topic status implementation"

This reverts commit b134dcec1b451365b2897ad93833af7de25b3b59.

guilherme-brandao committed:
Unify stake remove completed events

guilherme-brandao committed:
Added more ABCI related events (#853)

Added these events:

### **1. Stake Removal Events**
1. **`EmitReputerStakeRemovalCompletedEvent`**
   - Tracks when reputer stake removal is completed
   - Parameters: `topicId`, `blockHeight`, `reputer`, `amount`

2. **`EmitDelegateStakeRemovalCompletedEvent`** 
   - Tracks when delegate stake removal is completed
- Parameters: `topicId`, `blockHeight`, `delegator`, `reputer`, `amount`

### **2. Delegate Reward Events**
3. **`EmitDelegateRewardShareUpdatedEvent`**
   - Tracks when delegate reward share percentage is updated
   - Parameters: `topicId`, `reputer`, `rewardPerShare`, `blockHeight`

4. **`EmitDelegateRewardDistributedEvent`**
   - Tracks when delegate rewards are distributed to pending account
   - Parameters: `topicId`, `reputer`, `amount`, `blockHeight`

### **3. Active Participants Tracking Events**
5. **`EmitActiveReputersSetEvent`**
   - Tracks active reputers for a topic/nonce during loss calculation
   - Parameters: `topicId`, `nonceBlockHeight`, `addresses`

6. **`EmitActiveInferersSetEvent`**
- Tracks active inferers for a topic/nonce during worker nonce closing
   - Parameters: `topicId`, `nonceBlockHeight`, `addresses`

7. **`EmitActiveForecastersSetEvent`**
- Tracks active forecasters for a topic/nonce during worker nonce
closing
   - Parameters: `topicId`, `nonceBlockHeight`, `addresses`

8. **`EmitActiveTopicsAtBlockSetEvent`**
   - Tracks which topics are scheduled to be active at a future block
   - Parameters: `targetBlockHeight`, `topicIds`

### **4. Network Inference Weight Events**
9. **`EmitNetworkInferenceInfererWeightSetEvent`**
   - Tracks inferer weights during network inference calculation
   - Parameters: `topicId`, `nonceBlockHeight`, `address`, `weight`

10. **`EmitNetworkInferenceForecasterWeightSetEvent`**
    - Tracks forecaster weights during network inference calculation
    - Parameters: `topicId`, `nonceBlockHeight`, `address`, `weight`

### **5. Network Inference Regret Events**
11. **`EmitNetworkInferenceInfererRegretUsedEvent`**
- Tracks regret values used for inferers during network inference
calculation
    - Parameters: `topicId`, `nonceBlockHeight`, `address`, `regret`

12. **`EmitNetworkInferenceForecasterRegretUsedEvent`**
- Tracks regret values used for forecasters during network inference
calculation
    - Parameters: `topicId`, `nonceBlockHeight`, `address`, `regret`

### **6. Topic Weight Events**
13. **`EmitTopicWeightUpdatedEvent`**
    - Tracks when topic weights are updated
    - Parameters: `topicId`, `newWeight`

### **7. Submission Window Events**
14. **`EmitWorkerSubmissionWindowOpenedEvent`**
    - Tracks when worker submission windows are opened
- Parameters: `topicId`, `nonceBlockHeight`, `windowStartBlock`,
`windowEndBlock`

15. **`EmitWorkerSubmissionWindowClosedEvent`**
    - Tracks when worker submission windows are closed
    - Parameters: `topicId`, `nonceBlockHeight`

16. **`EmitReputerSubmissionWindowOpenedEvent`**
    - Tracks when reputer submission windows are opened
- Parameters: `topicId`, `nonceBlockHeight`, `windowStartBlock`,
`windowEndBlock`

17. **`EmitReputerSubmissionWindowClosedEvent`**
    - Tracks when reputer submission windows are closed
    - Parameters: `topicId`, `nonceBlockHeight`

guilherme-brandao committed:
Added reviewer suggestions

xmariachi committed:
Emit emission info event every block (#854)

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v           ✰  Thanks for creating a PR! You're awesome! ✰
v Please note that maintainers will only review those PRs with a
completed PR template.
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Purpose of Changes and their Description
* Emit an emission info event every block. 
* Refactor the keeper/queryserver so that query logic happens on the
keeper, and it can be used as a query, or in the beginblocker.

## Are these changes tested and documented?

- [x] If tested, please describe how. If not, why tests are not needed.
-- Added unit tests, although the tested code has just been refactored,
not changed.
- [x] If documented, please describe where. If not, describe why docs
are not needed.
- [x] Added to `Unreleased` section of `CHANGELOG.md`?

xmariachi committed:
fix failing test on order of attributes

xmariachi committed:
fix mint codec.go generation

xmariachi committed:
add changelog

xmariachi committed:
remove extra comment

guilherme-brandao committed:
Add new actor window queries (#855)

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v           ✰  Thanks for creating a PR! You're awesome! ✰
v Please note that maintainers will only review those PRs with a
completed PR template.
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Purpose of Changes and their Description

## Link(s) to Ticket(s) or Issue(s) resolved by this PR

## Are these changes tested and documented?

- [ ] If tested, please describe how. If not, why tests are not needed.
- [ ] If documented, please describe where. If not, describe why docs
are not needed.
- [ ] Added to `Unreleased` section of `CHANGELOG.md`?

## Still Left Todo

*Fill this out if this is a Draft PR so others can help.*

---------

Co-authored-by: Diego C <xmariachi@gmail.com>

amimart commented on x/emissions/proto/emissions/v9/query.proto:
```diff
@@ -555,6 +540,18 @@ service QueryService {
+    option (google.api.http).get = "/emissions/v9/worker_submission_window_status/{topic_id}";
+  }
+
+  // Get reputer submission window status and timing
+  rpc GetReputerSubmissionWindowStatus(GetReputerSubmissionWindowStatusRequest) returns (GetReputerSubmissionWindowStatusResponse) {
+    option (cosmos.query.v1.module_query_safe) = true;
+    option (google.api.http).get = "/emissions/v9/reputer_submission_window_status/{topic_id}";
+  }
```
We can have multiple nonce open at the same time so I think we should query the window status for a nonce, not only a topic.

Also maybe we could have a single query to provide both worker and reputer window information

xmariachi replied:
Discussed already, but for the record. 
1. Yes - the purpose of this query is tailored to actor usage, which usually wants to know the latest nonce to action upon, and any subsequent or next nonce. 
2. "a single query to provide both worker and reputer window information" = this makes complete sense, a query informing the lifecycle of a particular nonce (at which point the nonce is, or at which point broke - eg if there worker nonce closed but there were no inferences, then the reputer does not get created - this could be informed as the lifecycle for the nonce, provided the nonce is available). Created ticket about it.

xmariachi reviewed (approved):
all my concerns are tackled :)

amimart reviewed (approved):
🚀
